feat: crud на классах

alinkalina opened the pull request.

alinkalina committed:
feat: class based crud

alinkalina committed:
fix: remove type errors

LaGGgggg commented:
Это запоздалое review (кидаю его ещё и сюда, чтобы было удобно):

### `db` в `__init__`:
`db` используется в каждом методе `BaseCrud`. Так зачем передавать его в эти методы в качетсве аргумента? Можно же устанавливать `db` при инициализации класса, а потом получать к нему доступ в каждом методе через `self.db`. Таким образом мы упрощаем использование классов (и их методов), написание новых методов и улучшаем безопасность (чем проще, тем, зачастую безопаснее, так как сложно накосячить)
### ABC:
[ABC](https://docs.python.org/3/library/abc.html) превращает класс в абстрактный, то есть класс, предназначенный **только** наследования. Что даёт:
- Не можем создать экземпляр абстрактного класса (будет ошибка)
- Мы можем написать метод без реализации, который класс-наследник будет **обязан** реализовать (в примере ниже это не используется)
В нашем случае мы абстарактным классом показываем, что это базовый класс, который нельзя использовать напрямую, только через наследование.
[more](https://docs-python.ru/tutorial/klassy-jazyke-python/abstraktnye-klassy/)

```python
from abc import ABC


class BaseCrud(ABC):  
    ...
```

### `__init__`:

```python title:Было:
class BaseCrud:  
    ...
  
class UserCrud(BaseCrud):  
    def __init__(self):  
        self.model = models.User  

	def create(self, object_to_create: Type[schemas.BaseModel]) -> Type[Base]:  
	    db_object = self.model(**object_to_create.dict())  
  
	    self.add_to_db_and_refresh(db_object)  
  
	    return db_object
	...
```
```python title:"Почему плохо:"
class BadCrud(BaseCrud):  
    def __init__(self):  
        pass
	...
```
```python title:Стало:
class BaseCrud(ABC):  
    def __init__(self, model: Type[Base], db: Session) -> None:  
        self.model = model  
        self.db = db  
    ...

  
class UserCrud(BaseCrud):  
    def __init__(self, db: Session) -> None:  
        super().__init__(models.User, schemas.UserCreate, db)  
    ...
```
Про [super()](https://docs-python.ru/tutorial/vstroennye-funktsii-interpretatora-python/funktsija-super/)

### `*args` и `**kwargs`:
Что это всё значит? Представим себе функцию `f`, которая принимает два аргумента и возвращает их сумму:
```python
def f(a, b):
	return a + b
```
Всё круто, но что, если нам нужно сложить переменное количество чисел (то есть в одном случае 3, а в другом - 8)? А вот тут нам и пригодится `*args`:
```python
def f(*args):

	result = 0

	for n in args:
		result += n

	return result


f(2, 4, 6, 8)  # 20
f(1, 2)  # 3
f(9, 9, 9)  # 27
```
Теперь наша функция будет принимать переменное количество аргументов и складывать их в кортеж (tuple) `args`. `**kwargs` работает похожим образом, но предназначен для именованных аргументов (и складывает всё не в кортеж, а в словарь):
```python
def f(**kwargs):
	return kwargs


f(foo=7, bar=2)  # {'foo': 7, 'bar': 2}
```
Теперь мы можем применить наши знания в `BaseCrud`:
```python
class BaseCrud(ABC):  
    ...
    def get(self, *args, **kwargs) -> Type[Base]:  
        return self.db.query(self.model).get(*args, **kwargs)  
  
    def get_many(self, *args, **kwargs) -> list[Type[Base]]:  
        return self.db.query(self.model).filter(*args, **kwargs).all()
```
Воаля, теперь мы можем вызвать `get` в классе-наследнике с любыми аргументами. Теперь нам не нужно переписывать `get` для каждой новой модели, мы просто будем вызывать `get(username='username')` или `get(name='my')`, в зависимости от модели.
Статьи: [тык](https://tproger.ru/translations/python-args-and-kwargs), [тык](https://stackoverflow.com/questions/287085/what-do-args-and-kwargs-mean), [тык](https://habr.com/ru/companies/ruvds/articles/482464/)
P.S. Это всё **нужно** потыкать некоторое время, чтобы в голове уложилось как именно оно работает. И эта тема не заканчивается только на переменном количестве аргументов функций, можно сделать, например, так: `f(*my_list)`, тогда все элементы списка `my_list` будут переданы по отдельности. А ещё, необязательно использовать `args` и `kwargs`, можно поставить любое слово после `*` и `**`, это будет работать (но такое не приветствуется сообществом в большинстве случаев).

### `@staticmethod`:
```python title:Было
def get_password_hash(self, password: str, pwd_context: CryptContext) -> str:  
    return pwd_context.hash(password)
```
```python title:Стало
@staticmethod  
def get_password_hash(password: str, pwd_context: CryptContext) -> str:  
    return pwd_context.hash(password)
```
Зачем? При помощи `@staticmethod` мы говорим питону, что этот метод **не будет** использовать self. Фактически, это статичная функция, её можно без проблем вынести из класса, но мы оставляем её в классе (в данном случае это обусловлено семантически, этот метод относится именно к данной логике)
### Все изменения в одном месте (файл целиком):
***Тут не всё доделано, это только изменения, описанные ранее***
```python
from abc import ABC  
from typing import Type  
  
from sqlalchemy.orm import Session  
from passlib.context import CryptContext  
  
from core import schemas  
from . import models  
from .database import Base  
  
  
class BaseCrud(ABC):  
    def __init__(self, model: Type[Base], db: Session) -> None:  
        self.model = model  
        self.db = db  
  
    def add_to_db_and_refresh(self, object_to_add: Type[Base]) -> None:  
        self.db.add(object_to_add)  
        self.db.commit()  
        self.db.refresh(object_to_add)  
  
    def create(self, object_to_create: Type[schemas.BaseModel]) -> Type[Base]:  
        db_object = self.model(**object_to_create.dict())  
  
        self.add_to_db_and_refresh(db_object)  
  
        return db_object  
  
    def get(self, *args, **kwargs) -> Type[Base]:  
        return self.db.query(self.model).get(*args, **kwargs)  
  
    def get_many(self, *args, **kwargs) -> list[Type[Base]]:  
        return self.db.query(self.model).filter(*args, **kwargs).all()  
  
  
class UserCrud(BaseCrud):  
    def __init__(self, db: Session) -> None:  
        super().__init__(models.User, db)  
  
    def create_with_pwd_context(self, user: schemas.UserCreate, pwd_context: CryptContext):  
        hashed_password = self.get_password_hash(user.password, pwd_context)  
  
        db_user = self.model(username=user.username, hashed_password=hashed_password)  
  
        self.add_to_db_and_refresh(db_user)  
  
        return db_user  
  
    @staticmethod  
    def get_password_hash(password: str, pwd_context: CryptContext) -> str:  
        return pwd_context.hash(password)  
  
  
class ScopeCrud(BaseCrud):  
    def __init__(self):  
        self.model = models.Scope  
        self.scheme = schemas.ScopeCreate  
  
  
class UserToScopeCrud(BaseCrud):  
    def __init__(self):  
        self.model = models.UserToScope  
        self.scheme = schemas.UserToScopeCreate  
  
    def get_user_scopes(self, db: Session, user: schemas.User) -> list[models.Scope]:  
        return db.query(models.Scope).join(self.model).filter(self.model.user_id == user.id).all()
```

## Что доделать?
 - [x] С учётом нововедений, можно переписать метод `create_with_pwd_context` класса `UserCrud` на модификацию `create`. То есть мы переименовываем `create_with_pwd_context` в `create`, чуток подшаманиваем так, чтобы наш переименованный метод использовал уже готовый метод `create` базового класса. По сути, в нашем новом методе останется только логика создания `hashed_password` и схемы создания пользователя, остальное сделает исходный метод `create` базового класса. *(тут мб не очень понятно, могу пояснить дополнительно, если нужно)*
 - [x] Привести `ScopeCrud` и `UserToScopeCrud` в подходящий вид с учётом новой логики (по сути, нужно поменять `__init__` и убрать всё лишнее)
 - [x] Подогнать все кусочки кода, которые используют crud к работе по-новому (да, опять)))).

alinkalina committed:
fix: оптимизация метода create

Сделали BaseCrud абстрактным.
Добавили в __init__ db и pwd_context.
В коде изменили все места с использованием crud опираясь на внесённые изменения.

LaGGgggg committed:
fix: метод get в BaseCrud

Исправлен метод get:
  Убран *args.
  get заменён на filter_by (внутри метода).

LaGGgggg committed:
feat: улучшены безопасность и удобство методов crud-ов

API/sql/crud.py:
  BaseCrud:
    create: **kwargs заменен на schema, добавлена типизация.
  UserCrud:
    Удалён метод get_by_username (для упрощения логики, так как есть get с тем же функционалом).
    create: улучшена безопасность (за счёт использования схемы) и читабельность, добавлена типизация.

API/core/schemas.py: добавлена схема UserCreateDB.
API/routers/users.py/get_user: get_by_username заменён на get.
API/utils.py: get_by_username заменён на get.

LaGGgggg committed:
feat: улучшено логирование utils.py

LaGGgggg committed:
Merge branch 'dev' into dev-class-based-crud

LaGGgggg committed:
fix: исправлены ошибки совместимости кода

API/routers/users.py:
  Обновлено создание пользователя в соответствии с новой логикой crud.

API/utils.py: добавлена поддержка discord_id.
API/sql/crud.py: исправлена опечатка.

LaGGgggg commented:
Внёс некоторые изменения:
- Улучшил безопасность
- Сделал crud-ы чуть-чуть удобнее
- Исправил ошибки совместимости с #6

LaGGgggg reviewed (approved):
Считаю, что проделана хорошая работа, очень круто всё получилось

Roman821 reviewed (approved):
Очень круто, огромная молодчина!
Проделана крутая работа!)))
Одобряю
